### PR TITLE
[#15] Added healthy node rebalance

### LINKED ISSUE #15: Adapt sampler to changes in node health status
At the moment we're using the same `Sampler` with the same set of NeoFS nodes all the time even though nodes can become unavailable. It leads to some ugly logic in connection picking also where we retry several times (and can fail in the end!) even though we know exactly which nodes are not available (and sampler just shouldn't give them to us). We need sampler/pool to take that into account and rebalance node probabilities (unhealthy node effectively has zero probability).

### PR DESCRIPTION
**Pull request by KirillovDenis**
closes #15 

Signed-off-by: Denis Kirillov <denis@nspcc.ru>

### CODE REVIEW

File: pkg/pool/pool.go [hunk 1]
```diff
@@ -108,28 +108,65 @@ func newPool(ctx context.Context, options *BuilderOptions) (Pool, error) {
...
+				ok := true
+				defer wg.Done()
+				tctx, c := context.WithTimeout(ctx, options.NodeRequestTimeout)
+				defer c()
+				if _, err := netmap.EndpointInfo(tctx); err != nil {
+					ok = false
+					newWeights[i] = 0
+				}
+				if ok {
+					newWeights[i] = options.weights[i]
```
Comment (KirillovDenis) [inline pkg/pool/pool.go:133]: I'm not sure if this needs to be protected by mutex

File: pkg/pool/sampler_test.go [hunk 2]
```diff
@@ -38,3 +42,67 @@ func TestSamplerStability(t *testing.T) {
...
+func (n netmapMock) EndpointInfo(_ context.Context, _ ...client.CallOption) (*client.EndpointInfo, error) {
+	return nil, n.err
+}
+
+func TestHealthyReweight(t *testing.T) {
+	var (
+		weights           = []float64{0.9, 0.1}
+		names             = []string{"node0", "node1"}
+		rebalanceInterval = 100 * time.Millisecond
+		sleepInterval     = 1000 * time.Millisecond
```
Comment (KirillovDenis) [inline pkg/pool/sampler_test.go:69]: In rare cases the test may fail. What way can we avoid it?

File: pkg/pool/pool.go [hunk 3]
```diff
@@ -108,28 +108,65 @@ func newPool(ctx context.Context, options *BuilderOptions) (Pool, error) {
...
+		newWeights = adjustWeights(newWeights, options.weights)
+		source := rand.NewSource(time.Now().UnixNano())
+		p.lock.Lock()
+		p.sampler = NewSampler(newWeights, source)
+		p.lock.Unlock()
+		ticker.Reset(options.ClientRebalanceInterval)
+	}
+}
+
+func adjustWeights(weights, defaultWeights []float64) []float64 {
```
Comment (roman-khimov) [inline pkg/pool/pool.go:152]: `(*Builder).Build` has similar code, it should be reused.

File: pkg/pool/pool.go [hunk 4]
```diff
@@ -108,28 +108,65 @@ func newPool(ctx context.Context, options *BuilderOptions) (Pool, error) {
...
+	for _, weight := range weights {
+		sum += weight
+	}
+	if sum > 0 {
+		for i, weight := range weights {
+			weights[i] = weight / sum
+		}
+	} else {
+		for i := range weights {
+			weights[i] = defaultWeights[i]
```
Comment (roman-khimov) [inline pkg/pool/pool.go:163]: Well, everyone is dead here, so what's the point?

File: pkg/pool/pool.go [hunk 5]
```diff
@@ -108,28 +108,65 @@ func newPool(ctx context.Context, options *BuilderOptions) (Pool, error) {
...
+				p.clientPacks[i].healthy = ok
+				p.lock.Unlock()
+			}(i, cPack.client)
+		}
+		wg.Wait()
+
+		newWeights = adjustWeights(newWeights, options.weights)
+		source := rand.NewSource(time.Now().UnixNano())
+		p.lock.Lock()
+		p.sampler = NewSampler(newWeights, source)
```
Comment (roman-khimov) [inline pkg/pool/pool.go:146]: If weights are not changed (most of the time that's the case) we shouldn't be doing this.

File: pkg/pool/pool.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/pool/pool.go:133]: Should be OK.

File: pkg/pool/sampler_test.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/pool/sampler_test.go:69]: `require.Eventually`?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/pool/pool.go [hunk 6]
```diff
@@ -108,28 +101,67 @@ func newPool(ctx context.Context, options *BuilderOptions) (Pool, error) {
...
+func startRebalance(ctx context.Context, p *pool, options *BuilderOptions) {
+	ticker := time.NewTimer(options.ClientRebalanceInterval)
+	newWeights := make([]float64, len(options.weights))
+
+	for range ticker.C {
+		healthyChanged := false
+		wg := sync.WaitGroup{}
+		for i, cPack := range p.clientPacks {
+			go func(i int, netmap client.Netmap) {
+				wg.Add(1)
```
Comment (roman-khimov) [inline pkg/pool/pool.go:117]: That's racy, it should be done one line above. Think of the main thread spawning goroutines, but none of them starting to execute before `wg.Wait()`.

File: pkg/pool/pool.go [hunk 7]
```diff
@@ -108,28 +101,67 @@ func newPool(ctx context.Context, options *BuilderOptions) (Pool, error) {
...
+			source := rand.NewSource(time.Now().UnixNano())
+			p.lock.Lock()
+			p.sampler = NewSampler(newWeights, source)
+			p.lock.Unlock()
+		}
+		ticker.Reset(options.ClientRebalanceInterval)
+	}
+}
+
+func adjustWeights(weights []float64) []float64 {
```
Comment (roman-khimov) [inline pkg/pool/pool.go:151]: I'd not expect this function to change `weights`, although it clearly does that.

File: pkg/pool/sampler_test.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/pool/sampler_test.go]: Is it stable now?

File: pkg/pool/pool.go [hunk 8]
```diff
@@ -108,28 +101,67 @@ func newPool(ctx context.Context, options *BuilderOptions) (Pool, error) {
...
-	}()
+	go startRebalance(ctx, pool, options)
 	return pool, nil
 }
 
+func startRebalance(ctx context.Context, p *pool, options *BuilderOptions) {
+	ticker := time.NewTimer(options.ClientRebalanceInterval)
+	newWeights := make([]float64, len(options.weights))
+
+	for range ticker.C {
```
Comment (roman-khimov) [inline pkg/pool/pool.go:112]: Maybe we can move the body of this loop into some other function to simplify testing? Then you can deterministically call this function in various ways and always get predictable behaviour. Then we just suppose that periodically calling it is fine too.

File: pkg/pool/sampler_test.go (hunk 2 above)
Comment (KirillovDenis) [reply pkg/pool/sampler_test.go]: Seems to be yes

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**